cql3/restrictions/restriction.hh

```diff
@@ -49,15 +50,87 @@
 #include "cql3/term.hh"
 #include "cql3/statements/bound.hh"
 #include "index/secondary_index_manager.hh"
+#include "query-result-reader.hh"
 #include "types.hh"
 
 namespace cql3 {
 
+namespace selection {
+class selection;
+} // namespace selection
+
 namespace restrictions {
 
+class restriction;
+
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+/// Work-in-progress new restriction representation.  When complete, it will replace the old one.
+namespace wip {
+
+// The new representation exposes its member data publicly.  Operations on restrictions are
+// performed by free functions that take restrictions as parameters and use the visitor pattern to
+// specialize code for different kinds of restrictions.
+//
+// The most interesting class is wip::binary_operator below, which can represent both multi- and
+// single-column restrictions.  Instead of the old merge_with mechanism for combining restrictions,
+// we will simply add them into a conjunction expression that will be processed using visitors.
+
+class binary_operator;
+class conjunction;
+
+/// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
+using expression = std::variant<bool, conjunction, binary_operator>;
+
+/// A column, optionally subscripted by a term (eg, c1 or c2['abc']).
+struct column_value {
+    const column_definition* col;
+    ::shared_ptr<term> sub; ///< If present, this LHS is col[sub], otherwise just col.
```

Comment:
Why not subscription is an operator? Is it planned for later? Would it be possible to eventually do expr[expr]?

Reply:
> Why isn't subscription an operator?

Because then LHS would need to be expression, which is too general for this first step.  When `m[123]` is a binary_operator, then `m[123]>0` is a binary operator whose LHS is also a binary_operator.

> Is it planned for later?

Sure, if we can work out, eg, what it means to subscript a tuple.

> Would it be possible to eventually do expr[expr]?

Yes.

Reply:
ok.